### PR TITLE
Add github-add-comment version 0.3

### PR DESCRIPTION
**Pull request by vinamra28**
# Changes

As of now github-add-comment task supports input of comments only via
params.

This PR adds support of workspaces if in case somebody wants to provide
the comments via some file. One such use case can be:-
In Catlin CI, logs are produced after executing catlin command on
changed files which can contain multiple warnings and these warnings can
be ingored by the user if not posted on the github PR as the CI is
green. So all the logs would be stored in a file with proper formatting
and then the shared workspace can be passed to the github-add-comment
task along with the filename containing the formatted logs.

Signed-off-by: vinamra28 <vinjain@redhat.com>
# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality of task changed or new task added)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [X] Complies with [Catalog Organization TEP][TEP], see [example]. **Note** [An issue has been filed to automate this validation][validation]
  - [X] File path follows  `<kind>/<name>/<version>/name.yaml`
  - [X] Has `README.md` at `<kind>/<name>/<version>/README.md`
  - [X] Has mandatory `metadata.labels` - `app.kubernetes.io/version` the same as the `<version>` of the resource
  - [X] Has mandatory `metadata.annotations` `tekton.dev/pipelines.minVersion`
  - [X] mandatory `spec.description` follows the convention

          ```

          spec:
            description: >-
              one line summary of the resource

              Paragraph(s) to describe the resource.
          ```

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md)
for more details._

---

[TEP]: https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md
[example]: https://github.com/tektoncd/catalog/tree/master/task/git-clone/0.1
[validation]:  https://github.com/tektoncd/catalog/issues/413

### CODE REVIEW

**Comment by chmouel** [comment]
Don't we want optional workspace instead of optional params?

**Comment by vinamra28** [comment]
> Don't we want optional workspace instead of optional params?

I am very bad at python :disappointed: so don't know how to check whether workspace is bound or not :sweat_smile:. So I did a workaround by adding one more step that will check whether file is passed or it's a comment. In any of the case it will be moved to a file `comment_private.txt` and in python script I am going to read that file and post the file contents as a comment.

**Comment by sbwsg** [comment]
> how to check whether workspace is bound or not

In your script I think it would look like:

```python
if "$(workspaces.foo.bound)" == "true":
```

**Comment by vinamra28** [comment]
> > how to check whether workspace is bound or not
> 
> In your script I think it would look like:
> 
> ```python
> if "$(workspaces.foo.bound)" == "true":
> ```

ohh... I thought it would be handled other way. Thanks @sbwsg :+1:

File: task/github-add-comment/0.3/github-add-comment.yaml [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        # api url path /repos/foo/issues/202
+        api_url = "$(params.API_PATH_PREFIX)" + "/repos/" + "/".join(split_url[1:3]) + \
+                  "/issues/" + split_url[-1]
+
+        commentParamValue = "$(params.COMMENT_OR_FILE)"
+
+        # check if workspace is bound and parameter passed is a filename or not
+        if "$(workspaces.comment-file.bound)" == "true" and os.path.exists(commentParamValue):
+          comment_file = open(commentParamValue, "r")
+          commentParamValue = comment_file.read()
```
Comment (chmouel) [inline task/github-add-comment/0.3/github-add-comment.yaml:88]: nit: you can inline both
```suggestion
          commentParamValue = open(commentParamValue, "r").read()
```

File: task/github-add-comment/0.3/README.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+See GitHub's documentation on [Understanding scopes for OAuth Apps](https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/) to figure out what scopes you need to give to this token to add comment to an issue or a pull request.
+
+## Parameters
+
+- **GITHUB_HOST_URL:**: The GitHub host domain (_default:_ `api.github.com`)
+- **API_PATH_PREFIX:**: The GitHub Enterprise has a prefix for the API path. _e.g:_ `/api/v3`
+- **REQUEST_URL:**: The GitHub pull request or issue url, _e.g:_
+  `https://github.com/tektoncd/catalog/issues/46`
+- **COMMENT_OR_FILE:**: The actual comment to add or the filename containing comment to post. _e.g:_ `don't forget to eat your vegetables before commiting.` _or_ `input.txt`
```
Comment (chmouel) [inline task/github-add-comment/0.3/README.md:42]: see below comment too.

```suggestion
- **COMMENT_OR_FILE:**: The actual comment to add or a relative path to a filename inside the optional workspace `comment-file` containing comment to post. _e.g:_ `don't forget to eat your vegetables before commiting.` _or_ `input.txt`
```

File: task/github-add-comment/0.3/github-add-comment.yaml [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+        # This will convert https://github.com/foo/bar/pull/202 to
+        # api url path /repos/foo/issues/202
+        api_url = "$(params.API_PATH_PREFIX)" + "/repos/" + "/".join(split_url[1:3]) + \
+                  "/issues/" + split_url[-1]
+
+        commentParamValue = "$(params.COMMENT_OR_FILE)"
+
+        # check if workspace is bound and parameter passed is a filename or not
+        if "$(workspaces.comment-file.bound)" == "true" and os.path.exists(commentParamValue):
```
Comment (chmouel) [inline task/github-add-comment/0.3/github-add-comment.yaml:86]: I think we want to have COMMENt_OR_FILE relative to the file in the workspace? 

ie:


```suggestion
        if "$(workspaces.comment-file.bound)" == "true" and os.path.exists(os.path.join("$(workspaces.comment-file.path)", commentParamValue)):
```

need to update documentation too, see above comment earlier.

Comment (vinamra28) [reply task/github-add-comment/0.3/github-add-comment.yaml:86]: I have already set the `workingDir` as `workingDir: $(workspaces.comment-file.path)` so do I still need to update the following line?

Comment (chmouel) [reply task/github-add-comment/0.3/github-add-comment.yaml:86]: I am not so sure how it works with regard to workingDir and optional workspace, if the workspace is not there where the workingDir would be then ?

Comment (vinamra28) [reply task/github-add-comment/0.3/github-add-comment.yaml:86]: then the `workingDir` would be set to `/workspace`

Comment (vinamra28) [reply task/github-add-comment/0.3/github-add-comment.yaml:86]: so if the workspace is bound then `workingDir` would be set as `/workspace/comment-file` else `/workspace`

**Comment by chmouel** [comment]
/lgtm

thanks

**Comment by chmouel** [comment]
/lgtm
/approve

🥳

**Pull request merged**